modules/nodes-nodes-rebooting-gracefully.adoc

```diff
@@ -22,7 +22,29 @@ $ oc adm cordon <node1>
 +
 [source,terminal]
 ----
-$ oc adm drain <node1> --ignore-daemonsets --delete-emptydir-data
+$ oc adm drain <node1> --ignore-daemonsets --delete-emptydir-data --grace-period=300 --force
+----
+
+. (Optional) PodDisruptionBudget exists:
+
++ 
+If there is any PodDisruptionBudget defined on the cluster, some extra steps have to be performed before draining the node, otherwise the draining will fail.
+
++
+Either scale the deployment which has the PodDisruptionBudget defined to 0:
++
+[source,terminal]
+----
+$ oc scale deployment <deploy> -n <namespace> --replicas=0
+----
+
++
+or delete the PodDisruptionBudget temporarily:
+
++
+[source,terminal]
+----
+$ oc delete -f pdb.yaml 
```

Comment:
I don't like this at all, in particular the part about deleting the PDB. If someone put a PDB there, it's probably for a reason and we shouldn't mess with this from the infrastructure side.

If that PDB comes from one of our (Red Hat) components and is blocking a reboot, we should file a bug against that component.
If it's a 3:rd party application, we could add a section on recommended practices of using PDBs, something along the lines of https://kubernetes.io/docs/tasks/run-application/configure-pdb/#think-about-how-your-application-reacts-to-disruptions

Regarding scaling the deployment, I think this is more acceptable but we should not make any assumptions that the controlling entity is a Deployment. Instead, I suggest we should make this more generic and state that the application must be undeployed if can't support an in-service update of the cluster.

Reply:
Thanks @djuran for the comments, when I did the test on 4.9/4.8, there was no PDBs came from OpenShift itself. To delete PDBs or scale down the deployment was more about the custom PDBs. I agree with you if any PDB from OpenShift blocks the draining on SNO, OpenShift shall fix the issue, either bypassing the check or specifying from command line parameter like '--skip-pdbs'(not exist now). But those needs the code change.

Since @mburke5678 re-organized the original PR in http://file.rdu.redhat.com/~mburke/update-node-reboot-procedure/nodes/nodes/nodes-nodes-rebooting.html#nodes-nodes-rebooting-gracefully_nodes-nodes-rebooting, @mburke5678 can you please review David's comments about PDBs part to find a better place and desciption in the docs? I can definitely help as well, Thanks.

Reply:
Does `--disable-eviction=false` work in this case? 

`Force drain to use delete, even if eviction is supported. This will bypass checking PodDisruptionBudgets, use with caution.`

Reply:
Hi @mburke5678 , that might be a better way, sorry I was not aware of this parameter before. @djuran with this option we can avoid touching PDBs and deployment, how do you think? 

@mburke5678 can you share us a preview link so that we can dry-run one more time? Thanks.

Reply:
Sorry, should have replied in the thread instead of new comment.... But what is the purpose of cordoning and draining an SNO in the first place? It's not like the Pods have anywhere else to go...

Reply:
@borball Sorry. I didn't see your request for a preview. 
http://file.rdu.redhat.com/~mburke/update-node-reboot-procedure/nodes/nodes/nodes-nodes-rebooting.html#nodes-nodes-rebooting-gracefully_nodes-nodes-rebooting

Reply:
Sorry @mburke5678 for replying back late.
According to my test '--disable-eviction' will ignore the pdbs but '--disable-eviction=false' won't, following is the related error: 
`error when evicting pods/"nginx-747fbd9b9b-7bjzl" -n "pdb-test" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.`

So we should update the description a bit:
- (default)For a normal case where has no pdb defined, use command 
  `oc adm drain <node1> --ignore-daemonsets --delete-emptydir-data --force --grace-period=60`
- In case there is any pdbs defined and blocks the drain, run command
  `oc adm drain <node1> --ignore-daemonsets --delete-emptydir-data --force --disable-eviction --grace-period=60`

The customer can choose to leave the pods which has pdbs defined in an incompleted-evicted state as well since the node will be drained in the end, only the pods which has pdbs are remaining and not fully evicted. Like the case below:

```
[root@hub-helper sno147]# oc get nodes
NAME                          STATUS                     ROLES           AGE   VERSION
sno147.outbound.vz.bos2.lab   Ready,SchedulingDisabled   master,worker   8d    v1.22.3+2cb6068
[root@hub-helper sno147]# oc get pdb
NAME        MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
nginx-pdb   3               N/A               0                     20m
[root@hub-helper sno147]# 
[root@hub-helper sno147]# 
[root@hub-helper sno147]# oc get pods
NAME                     READY   STATUS    RESTARTS   AGE
nginx-747fbd9b9b-2m2r5   0/1     Pending   0          55s
nginx-747fbd9b9b-78q54   1/1     Running   0          5m47s
nginx-747fbd9b9b-bnfj6   1/1     Running   0          8m4s
nginx-747fbd9b9b-gr9tz   1/1     Running   0          8m4s
nginx-747fbd9b9b-m8kzd   0/1     Pending   0          44s
```

Reply:
@borball @djuran Please see my latest updates. 
[Preview](http://file.rdu.redhat.com/~mburke/update-node-reboot-procedure/nodes/nodes/nodes-nodes-rebooting.html#nodes-nodes-rebooting-gracefully_nodes-nodes-rebooting)  
[commit1](https://github.com/openshift/openshift-docs/pull/46090/commits/f200c91bd40cd0f1e27436dd622f98bb938a596b) 
[commit2](https://github.com/openshift/openshift-docs/pull/46090/commits/c40184eac9e608d6ee74d2cc59f147ec9ddb0b59) 
* Added to the first note a sentence about [why to cordon and drain](https://github.com/openshift/openshift-docs/pull/43659#discussion_r891196563) when the pods don't [have anywhere else to go](https://github.com/openshift/openshift-docs/pull/43659#discussion_r891049275).
* Added [two commands for clusters with and without custom PDBs](https://github.com/openshift/openshift-docs/pull/43659#discussion_r896001926), removing the `=false` from `--disable-eviction`. 
* Removed _For  Single Node OpenShift (SNO) clusters_ [as requested by djuran](https://github.com/openshift/openshift-docs/pull/46090#discussion_r891046934).

Reply:
Thanks for updating this @mburke5678, It looks good to me now.

Reply:
@mburke5678 Regarding the 4.6, SNO was not avaialble in 4.6, so the text might be updated to hide SNO related info in that case.